### PR TITLE
Added support for PHP 8.0 promoted parameters

### PR DESCRIPTION
**Pull request by michaelpetri**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | yes
| QA            | no

### Description

Adds support for promoted parameters

Solves (partially) https://github.com/laminas/laminas-code/issues/104

### CODE REVIEW

File: test/Generator/ClassGeneratorTest.php [hunk 1]
```diff
@@ -1313,4 +1321,98 @@ public function testCorrectImplementNames()
...
+        eval('namespace ' . __NAMESPACE__ . '; class ' . $className . '{ public function __construct( public string \$inInitializer){} }');
+
+        $classGenerator = ClassGenerator::fromReflection(new ReflectionClass(__NAMESPACE__ . '\\' . $className));
+
+        $expectedOutput = <<<EOS
+class ClassWithInInitializer
+{
+    public function __construct(
+        public string $inInitializer
+    );
```
Comment (Ocramius) [inline test/Generator/ClassGeneratorTest.php]: `{}` still needed

File: test/Generator/ClassGeneratorTest.php [hunk 2]
```diff
@@ -1313,4 +1321,98 @@ public function testCorrectImplementNames()
...
+                )
+            )
+        ]);
+
+        $expectedOutput = <<<EOS
+class InInitializerWithDefaultValue
+{
+    public function __construct(
+        private Foo \$inInitializer = Foo::bar(baz: 'qux')
+    );
```
Comment (Ocramius) [inline test/Generator/ClassGeneratorTest.php]: `{}` still needed

File: test/Generator/ClassGeneratorTest.php [hunk 3]
```diff
@@ -1313,4 +1321,98 @@ public function testCorrectImplementNames()
...
+        $classGenerator->addMethod('__construct', [
+            new InInitializerGenerator('inInitializer', 'string')
+        ]);
+
+        $expectedOutput = <<<EOS
+class InInitializer
+{
+    public function __construct(
+        public string \$inInitializer
+    );
```
Comment (Ocramius) [inline test/Generator/ClassGeneratorTest.php]: `{}` still needed

File: test/Generator/ClassGeneratorTest.php [hunk 4]
```diff
@@ -1313,4 +1321,98 @@ public function testCorrectImplementNames()
...
+
+        self::assertEquals($expectedOutput, $classGenerator->generate());
+    }
+
+    /** @requires PHP >= 8.1 */
+    public function testClassFromReflectionWithInInitializer(): void
+    {
+        $className = uniqid('ClassWithInInitializer', false);
+
+        eval('namespace ' . __NAMESPACE__ . '; class ' . $className . '{ public function __construct( public string \$inInitializer){} }');
```
Comment (Ocramius) [inline test/Generator/ClassGeneratorTest.php]: `\` not needed here

File: src/Generator/ObjectGenerator.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\Code\Generator;
+
+final class ObjectGenerator
```
Comment (Ocramius) [inline src/Generator/ObjectGenerator.php]: Discussed in private call: we will not implement this for now

Comment (Ocramius) [reply src/Generator/ObjectGenerator.php]: We'll need a new issue for this - will create one after clearing up ideas about this, for now, and for now, we will raise an exception in `::fromReflection()`, when an `object` as default value is encountered (something like "unsupported, for now")

**Review by Ocramius — CHANGES_REQUESTED** [review]
Discussed this review in a 1-on-1 session: PHP 8.1 is currently a bit out of reach.


```php
<?php

interface Bar {}
class Baz implements Bar {}
class Foo
{
    public function __construct(
        public Bar $bar = new Baz()
    ) {}
    
    public function doThing(Bar $bar = new Baz()) {
        return $bar;
    }
}

var_dump(new Foo());

var_dump((new Foo)->doThing());
```

https://3v4l.org/SoOPU

In practice, we should stop this patch at only supporting initializers (and visibility thereof), while we will skip `new_in_initializers` for now ( https://wiki.php.net/rfc/new_in_initializers ) because reflection for it does not give us enough information about the default value yet.

We will likely need AST-based reflector for that, and we're far from that: at that point, this library will either need phasing out, or a full rewrite.

File: src/Generator/ClassGenerator.php [hunk 6]
```diff
@@ -140,7 +141,17 @@ public static function fromReflection(ClassReflection $classReflection)
             }
 
             if ($reflectionMethod->getDeclaringClass()->getName() == $className) {
-                $methods[] = MethodGenerator::fromReflection($reflectionMethod);
+                $method = MethodGenerator::fromReflection($reflectionMethod);
+
+                if (self::CONSTRUCTOR_NAME === $method->getName()) {
```
Comment (michaelpetri) [inline src/Generator/ClassGenerator.php]: Should probably better moved to `MethodGenerator::isConstructor()`

Comment (Ocramius) [reply src/Generator/ClassGenerator.php]: Would be fine to keep it as-is for now, IMO. Just `strtolower()` needed

Comment (michaelpetri) [reply src/Generator/ClassGenerator.php]: Added

**Review by Ocramius — APPROVED** [review]
LGTM, excellent work, @michaelpetri!

**Pull request merged**